File: backend/app/models/asset.py
from datetime import datetime
from typing import Any

from sqlalchemy import BigInteger, DateTime, ForeignKey, Index, String, Text, text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class Asset(Base):
    __tablename__ = "assets"
    __table_args__ = (
        Index("idx_assets_project_floor", "project_id", "floor_id"),
        Index("idx_assets_uploaded_by", "uploaded_by"),
    )

    id: Mapped[str] = mapped_column(
        UUID(as_uuid=False),
        primary_key=True,
        server_default=text("gen_random_uuid()"),
    )
    project_id: Mapped[str] = mapped_column(
        UUID(as_uuid=False),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
    )
    floor_id: Mapped[str | None] = mapped_column(
        UUID(as_uuid=False),
        ForeignKey("floors.id", ondelete="SET NULL"),
        nullable=True,
    )
    uploaded_by: Mapped[str | None] = mapped_column(
        UUID(as_uuid=False),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    asset_type: Mapped[str] = mapped_column(String(50), nullable=False)
    source_format: Mapped[str | None] = mapped_column(String(50), nullable=True)
    storage_url: Mapped[str] = mapped_column(Text(), nullable=False)
    mime_type: Mapped[str | None] = mapped_column(String(100), nullable=True)
    file_size_bytes: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    metadata_json: Mapped[dict[str, Any]] = mapped_column(
        JSONB,
        nullable=False,
        server_default=text("'{}'::jsonb"),
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=text("now()")
    )

    project = relationship("Project", back_populates="assets")
    floor = relationship("Floor", back_populates="assets")
    uploader = relationship("User", back_populates="uploaded_assets")
    source_scene_drafts = relationship("SceneDraft", back_populates="source_asset")

File: frontend/src/types/asset.ts
import type { ISODateString, UUID } from './common';

export type AssetType = 'floorplan' | 'photo' | 'document' | string;

// §5.1 응답 (POST /floors/{floor_id}/assets) 및 §5.3 GET /assets/{id}
export interface Asset {
  id: UUID;
  floor_id: UUID | null;
  uploaded_by: UUID | null;
  asset_type: AssetType;
  source_format: string | null;
  storage_url: string;
  mime_type: string | null;
  file_size_bytes: number | null;
  metadata_json: Record<string, unknown>;
  created_at: ISODateString;
}

export interface UploadAssetParams {
  file: File;
  asset_type: AssetType;
}
